File: src/main/java/com/frightanic/smn/api/SmnData.java
package com.frightanic.smn.api;

import com.google.common.base.Splitter;
import com.wordnik.swagger.annotations.ApiModel;

import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Represents the entire data set for a point in time.
 */
@ApiModel(value = "A container for SMN data records")
public class SmnData {
  private final Map<String, SmnRecord> smnRecords;

  public SmnData(String sourceData) {
    List<String> records = Splitter.on("\n").trimResults().omitEmptyStrings().splitToList(sourceData);
    smnRecords = new HashMap<>(records.size() - 2);
    // omit first three lines (the empty one was already removed be the Splitter, omitEmptyStrings)
    /*
     * MeteoSchweiz / MeteoSuisse / MeteoSvizzera / MeteoSwiss
     *
     * stn;time;tre200s0;rre150z0;sre000z0;gre000z0;ure200s0;tde200s0;dkl010z0;fu3010z0;fu3010z1;prestas0;pp0qffs0;pp0qnhs0;ppz850s0;ppz700s0;dv1towz0;fu3towz0;fu3towz1;ta1tows0;uretows0;tdetows0
     */
    for (int i = 2; i < records.size(); i++) {
      SmnRecord smnRecord = new SmnRecord(records.get(i));
      smnRecords.put(smnRecord.getCode(), smnRecord);
    }
  }

  public Collection<SmnRecord> getAllRecords() {
    return smnRecords.values();
  }

  public SmnRecord getRecordFor(String code) {
    return smnRecords.get(code);
  }
}

File: src/main/kotlin/com/frightanic/smn/api/SmnData.kt
package com.frightanic.smn.api

/**
 * Represents the entire data set for a point in time.
 */
class SmnData(sourceData: String) {
    private val smnRecords = sourceData
        .split("\n")
        .map { it.trim() }
        .filter {
            it.isNotEmpty() &&
            // Skip any non-data line (comments, column headers, etc.).
            // Line has to start with the 3-char station code.
            // Example below:
            /*
             * Station/Location;Date;tre200s0;rre150z0;sre000z0;gre000z0;ure200s0;tde200s0;dkl010z0;fu3010z0;fu3010z1;prestas0;pp0qffs0;pp0qnhs0;ppz850s0;ppz700s0;dv1towz0;fu3towz0;fu3towz1;ta1tows0;uretows0;tdetows0
             * TAE;202208280630;16.50;0.00;0.00;41.00;98.90;16.30;22.00;2.90;5.40;954.10;1016.10;1017.40;-;-;-;-;-;-;-;-
             * COM;202208280630;14.50;0.00;0.00;27.00;84.60;11.90;329.00;5.40;7.90;948.70;1015.20;1016.20;-;-;-;-;-;-;-;-
             */
            it.matches(Regex("^[A-Z]{3};.*\$"))
        }
        .associate {
            val record = SmnRecord(it)
            record.code to record
        }

    fun getAllRecords() = smnRecords.values

    fun getRecordFor(code: String)= smnRecords[code]
}
